File: Java/src/com/leetcode/Easy/RankTransformOfAnArray.java
package com.leetcode.Easy;

import java.util.*;

// 1331. Rank Transform of an Array
public class RankTransformOfAnArray {
    public int[] arrayRankTransform(int[] arr) {
        int[] result = Arrays.copyOf(arr, arr.length);
        Arrays.sort(result);
        HashMap<Integer, Integer> map = new HashMap<>();
        for (int a : result)
            map.putIfAbsent(a, map.size() + 1);

        for (int i = 0; i < arr.length; ++i)
            result[i] = map.get(arr[i]);

        return result;
    }
}

File: JavaScript/Easy/RankTransformOfAnArray.js
/**
 * 1331. Rank Transform of an Array
 * @param {number[]} arr
 * @return {number[]}
 */
const arrayRankTransform = function (arr) {
  const sortedArr = [...arr].sort((a, b) => a - b);

  let rank = 1;
  let rankMap = new Map();
  for (let i = 0; i < sortedArr.length; i++) {
    if (!rankMap.has(sortedArr[i])) {
      rankMap.set(sortedArr[i], rank++);
    }
  }

  return arr.map(num => rankMap.get(num));
};